parsl/tests/test_monitoring/test_incomplete_futures.py

```diff
@@ -0,0 +1,68 @@
+import logging
+import os
+import parsl
+import pytest
+
+from concurrent.futures import Future
+
+logger = logging.getLogger(__name__)
+
+
+@parsl.python_app
+def this_app(inputs=()):
+    return inputs[0]
+
+
+@pytest.mark.local
+def test_future_representation():
+    import sqlalchemy
+    from sqlalchemy import text
+    from parsl.tests.configs.htex_local_alternate import fresh_config
+
+    if os.path.exists("runinfo/monitoring.db"):
+        logger.info("Monitoring database already exists - deleting")
+        os.remove("runinfo/monitoring.db")
+
+    logger.info("loading parsl")
+    parsl.load(fresh_config())
+
+    # this is arbitrary
+    TOKEN = 54739
+
+    # make a Future that has no result yet
+    # invoke a task that depends on it
+    # inspect and insert something about the monitoring recorded value of that Future
+    # make the Future complete
+    # inspect and insert something about the monitoring recorded value of that Future
+
+    # potential race condition:
+    # monitoring will take some time to update the record after submitting the task or completing the Future
+
+    f1 = Future()
+
+    f2 = this_app(inputs=[f1])
+
+    f1.set_result(TOKEN)
+
+    assert f2.result() == TOKEN
```

Comment:
Is there any value in checking the intermediate state of the monitoring DB between these statements?

Reply:
it's a bit non-deterministic so I didn't go that way - after line 43, the task row will appear (with an in-progress value) "eventually" in the monitoring database, but it's not synchronous with line 43 completing.

What I would expect is that various records would appear - basically the asserts that are in eg. PR #3056 - but I think it would be some code that has to sit there and poll the DB a few times until the "eventually consistent" DB reflects that the task is launched.

The interesting thing to check for the specific bug in this PR would be to check that the input field does not have `TOKEN` stored in it until after line 45, as a pre-condition.

That's also what could happen instead of shutting down the DFK with `cleanup` a few lines later: don't shut down parsl here, but instead poll the DB until it looks right.

I don't really have any strong feelings about whether it's the right thing to do or not.